openvm/src/extraction_utils.rs

```diff
@@ -0,0 +1,245 @@
+use std::collections::BTreeMap;
+use std::{collections::HashSet, sync::Arc};
+
+use crate::air_builder::AirKeygenBuilder;
+use crate::{BabyBearSC, IntoOpenVm};
+use itertools::Itertools;
+use openvm_circuit::arch::{VmChipComplex, VmConfig};
+use openvm_circuit_primitives::bitwise_op_lookup::SharedBitwiseOperationLookupChip;
+use openvm_circuit_primitives::range_tuple::SharedRangeTupleCheckerChip;
+use openvm_instructions::VmOpcode;
+use openvm_sdk::config::{SdkVmConfig, SdkVmConfigExecutor, SdkVmConfigPeriphery};
+use openvm_stark_backend::{
+    air_builders::symbolic::SymbolicConstraints, config::StarkGenericConfig, rap::AnyRap, Chip,
+};
+use openvm_stark_sdk::config::{
+    baby_bear_poseidon2::{config_from_perm, default_perm},
+    fri_params::SecurityParameters,
+};
+use openvm_stark_sdk::openvm_stark_backend::p3_field::PrimeField32;
+use openvm_stark_sdk::p3_baby_bear;
+use powdr_autoprecompiles::bus_map::{BusMap, BusType};
+use powdr_autoprecompiles::SymbolicMachine;
+
+use crate::utils::get_pil;
+
+use crate::customize_exe::openvm_bus_interaction_to_powdr;
+use crate::utils::symbolic_to_algebraic;
+
+fn to_option<T>(mut v: Vec<T>) -> Option<T> {
+    match v.len() {
+        0 => None,
+        1 => Some(v.pop().unwrap()),
+        _ => panic!("Expected at most one element, got multiple"),
+    }
+}
+
+fn get_bus_map<F: PrimeField32>(
+    chip_complex: &VmChipComplex<F, SdkVmConfigExecutor<F>, SdkVmConfigPeriphery<F>>,
+) -> BusMap {
+    let builder = chip_complex.inventory_builder();
+
+    let shared_bitwise_lookup =
+        to_option(builder.find_chip::<SharedBitwiseOperationLookupChip<8>>());
+    let shared_range_tuple_checker =
+        to_option(builder.find_chip::<SharedRangeTupleCheckerChip<2>>());
+
+    BusMap::from_id_type_pairs(
+        {
+            let base = &chip_complex.base;
+            [
+                (base.execution_bus().inner.index, BusType::ExecutionBridge),
+                (base.memory_bus().inner.index, BusType::Memory),
+                (base.program_bus().inner.index, BusType::PcLookup),
+                (
+                    base.range_checker_bus().inner.index,
+                    BusType::VariableRangeChecker,
+                ),
+            ]
+            .into_iter()
+        }
+        .chain(
+            shared_bitwise_lookup
+                .into_iter()
+                .map(|chip| (chip.bus().inner.index, BusType::BitwiseLookup)),
+        )
+        .chain(
+            shared_range_tuple_checker
+                .into_iter()
+                .map(|chip| (chip.bus().inner.index, BusType::TupleRangeChecker)),
+        )
+        .map(|(id, bus_type)| (id as u64, bus_type)),
+    )
+}
+
+pub fn get_airs_and_bus_map<P: IntoOpenVm>(
```

Comment:
I think the only reason this function does two things is because we want to use the same chip complex, right? But then this function could just take a reference to the chip complex, like `get_bus_map`?

Reply:
Agreed, but passing the complex exposes implementation details. We could either introduce:

```
struct OpenVmDriver {
   complex: VmChipComplex
}

impl OpenVmDriver {
   fn new(config: SdkVmConfig) -> Self {

   }

   fn bus_map(&self) -> BusMap

   fn airs(&self) -> Airs
}
```

And use cache if needed.

Or we have a single function which returns everything about the VM.

The latter is better if possible.

Could this be a follow up PR?

Reply:
Sure!